src/component/autocomplete.js

```diff
@@ -0,0 +1,312 @@
+    let eventFunction = event => {
+      if (this.keyCodes[keyName] === event.keyCode || keyName === '*') {
+        eventCallback(event);
+      }
+    };
+
+    this.listeners.push({
+      element  : element,
+      callback : eventCallback,
+      eventName: 'keydown'
+    });
+
+    element.addEventListener('keydown', eventFunction);
+  }
+
+  detached() {
+    this.removeEventListeners(this.listeners);
+  }
+
+  /**
+   * removes event listeners from DOM
+   *
+   * @param {Object[]} listeners objects that represent a event listener
+   */
+  removeEventListeners(listeners) {
+    listeners.forEach(listener => {
+      listener.element.removeEventListener(listener.eventName, listener.callback);
+    });
+  }
+
+  /**
+   * returns HTML that wraps matching substrings with strong tags
+   *
+   * @param {Object} result
+   *
+   * @returns {String}
+   */
+  labelWithMatches(result) {
+    return this.label(result).replace(
+      new RegExp(this.search, 'gi'),
+      match => {
+        return `<strong>${match}</strong>`;
+      });
+  }
+
+  /**
+   * Prepares the DOM by adding event listeners
+   */
+  attached() {
+    this.inputElement = this.element.querySelectorAll('input')[0];
+    this.dropdownElement = this.element.querySelectorAll('#dropdown')[0];
+
+    let openDropdown = (show = true) => {
+      this.showResults = show;
+    };
+
+    this.inputElement.addEventListener('focus', openDropdown);
+    this.inputElement.addEventListener('click', openDropdown);
+
+    this.inputElement.addEventListener('blur', () => openDropdown(false));
+
+    this.registerKeyDown(this.inputElement, 'down', event => {
+      this.selected = this.nextFoundResult(this.selected);
+      // do not move the cursor
+      event.preventDefault();
+    });
+
+    this.registerKeyDown(this.inputElement, 'up', event => {
+      this.selected = this.nextFoundResult(this.selected, true);
+      // do not move the cursor
+      event.preventDefault();
+    });
+
+    this.registerKeyDown(this.inputElement, 'enter', () => this.onSelect());
+
+    // tab closes the dropdown and jumps to the next tabable element (default behavior)
+    this.registerKeyDown(this.inputElement, 'tab', () => this.onSelect());
+  }
+
+  /**
+   * @param {Object} current selected item
+   * @param {Boolean} reversed when true gets the previous instead
+   *
+   * @returns {Object} the next of previous item
+   */
+  nextFoundResult(current, reversed) {
+    let index = (this.results.indexOf(current) + (reversed ? -1 : 1)) % (this.results.length);
+
+    if (index < 0) {
+      index = this.results.length - 1;
+    }
+
+    return this.results[index];
+  }
+
+  /**
+   * Set the text in the input to that of the selected item and set the
+   * selected item as the value. Then hide the results(dropdown)
+   *
+   * @param {Object} [result] when defined uses the result instead of the
+   * this.selected value
+   */
+  onSelect(result) {
+    if (result) {
+      this.selected = result;
```

Comment:
I would not want to override user specified values. If the value is `0`, `[]` `''` or `undefined` is the value then I would not want to default it to null